### PR TITLE
Fix/improve dirty regions on GBM/Wayland/X11 (EGL)

### PR DESCRIPTION
**Pull request by sarbes**
## Description
The previous dirty region implementation was inconsistent across the different platforms.

For X11, we just guessed how old the buffer we draw to is.

For GBM and Wayland, we guessed as well. But instead of just writing to the buffer, the GPU driver does a copy of the previous (N-1) buffer, because we specified the surface with the attribute "EGL_SWAP_BEHAVIOR" set to "EGL_BUFFER_PRESERVED".

With the patch, we can query the exact buffer age via "EGL_EXT_buffer_age". Even better, if "EGL_KHR_partial_update" is available, we can give the damaged region to the GPU driver, which in turn can omit untouched regions from rendering. The latter is only implemented on GBM/Wayland, but could be added to X11 (but I don't see much value in doing that).

If we can't query the buffer age, there are default values to fall back to ("2" for triple buffering, "3" on IOS). 

"EGL_KHR_partial_update" based dirty region rendering is only available when providing the following settings file ("algorithmdirtyregions" can be 1 or 2):
```
<advancedsettings>
  <gui>
    <algorithmdirtyregions>1</algorithmdirtyregions>
  </gui>
</advancedsettings>
```

## Motivation and context
Depending on the GPU, its driver and the windowing system; there might be some unnecessary copying going on. This patch might reduce GPU load.

Especially tile based GPUs which are using "EGL_KHR_partial_update" might see an performance uplift in certain situations.

## How has this been tested?
On my RX570 on X11, all dirty region modes are working fine.

On an Odroid C2 running LibreElec 11, it seems to run fine now.

## What is the effect on users?
More performance. The GPU only has to render updated parts. Less memory bandwidth consumed in certain situations. The device might run a bit cooler.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/utils/EGLUtils.cpp
Comment (neo1973) [inline xbmc/utils/EGLUtils.cpp:636]: This could be a little bit faster but I can't test it lacking a GPU with `EGL_KHR_partial_update`.
```suggestion
  using Rect = std::array<EGLint, 4>;

  if (dirtyRegions.empty())
  {
    static Rect zeroRect{};
    m_eglSetDamageRegionKHR(m_eglDisplay, m_eglSurface, zeroRect.data(), zeroRect.size());
  }
  else
  {
    EGLint height = CServiceBroker().GetWinSystem()->GetHeight();
    std::vector<Rect> rects;
    rects.reserve(dirtyRegions.size());
    for (const auto& i : dirtyRegions)
    {
      if (i.IsEmpty())
        continue;
      rects.push_back({(EGLint)i.x1, height - (EGLint)i.y2, (EGLint)i.Width(), (EGLint)i.Height()});
    }
    m_eglSetDamageRegionKHR(m_eglDisplay, m_eglSurface, reinterpret_cast<EGLint*>(rects.data()),
                            rects.size());
  }
```

File: xbmc/utils/EGLUtils.cpp
Comment (neo1973) [inline xbmc/utils/EGLUtils.cpp:607]: Is it necessary to query the extensions every time or can this be cached? E.g. in a static variable.

**Review by neo1973 — COMMENTED** [review]
Sounds like a good improvement :slightly_smiling_face:

File: xbmc/utils/EGLUtils.cpp
Comment (sarbes) [reply xbmc/utils/EGLUtils.cpp:607]: Currently, this is normal practice all over the place.

That's something I have on my agenda for GL/EGL extensions. I was thinking about doing the string compare at startup, so that we only have to query an enum. But I feel that this is something for another day.

File: xbmc/utils/EGLUtils.cpp
Comment (sarbes) [reply xbmc/utils/EGLUtils.cpp:636]: I'll try it out when I have the time.

**Comment by sarbes** [comment]
@ilmich this might interest you as well. Depending on the scene composition, it can reduce GPU and RAM load.

File: xbmc/windowing/WinSystem.h
Comment (lrusak) [inline xbmc/windowing/WinSystem.h:194]: Why int32_t? Why not just int or unsigned int?

File: xbmc/utils/EGLUtils.cpp
Comment (lrusak) [inline xbmc/utils/EGLUtils.cpp:191]: CEGLUtils has a helper for this

File: xbmc/utils/EGLUtils.cpp
Comment (lrusak) [inline xbmc/utils/EGLUtils.cpp:630]: C++ cast, same thing elsewhere.

File: xbmc/guilib/DirtyRegionTracker.h
Comment (lrusak) [inline xbmc/guilib/DirtyRegionTracker.h:16]: Remove explicit keyword

File: xbmc/application/Application.cpp
Comment (lrusak) [inline xbmc/application/Application.cpp:937]: This now seems to execute after the flip, should it be renamed BeforeRender?

Comment (sarbes) [reply xbmc/application/Application.cpp:937]: I moved it there because I can only query the buffer age after the flip. Maybe "AfterFlip"?

File: xbmc/windowing/WinSystem.h
Comment (sarbes) [reply xbmc/windowing/WinSystem.h:194]: Not particular reason. The query for GL returns an EGLInt, which is an int32_t as far as I know. Plus I had the idea to pass -1 for an unknown age.

What do you want me change it to?

File: xbmc/utils/EGLUtils.cpp
Comment (sarbes) [reply xbmc/utils/EGLUtils.cpp:630]: Roger.

File: xbmc/guilib/DirtyRegionTracker.h
Comment (sarbes) [reply xbmc/guilib/DirtyRegionTracker.h:16]: Roger.

File: xbmc/utils/EGLUtils.cpp
Comment (sarbes) [reply xbmc/utils/EGLUtils.cpp:191]: D'oh

**Comment by ilmich** [comment]
Hi sarbes,

I tried it yesterday, but I noticed some glitches in switching between video playback and the home page.
I don't know if it could depend on the fact that I use Nexus (patch apply without modifications), or the mesa version I use (23.0.0)
However, my gpu is a mali 400.

File: xbmc/windowing/WinSystem.h
Comment (garbear) [reply xbmc/windowing/WinSystem.h:194]: Using a sized int is making an assertion on the maximum value. Here, you're saying that the buffer age is < 2^31, but if it can only be 1 or 2, the assertion doesn't make sense, so I would use `unsigned int` to show that the bitsize isn't very important to the meaning.

File: xbmc/windowing/wayland/WinSystemWaylandEGLContext.h
Comment (yol) [inline xbmc/windowing/wayland/WinSystemWaylandEGLContext.h:42]:
```suggestion
  }
```

**Review by yol — COMMENTED** [review]
just glanced through the changes, wayland stuff seems OK!

**Comment by ninja-** [comment]
🤔

**Comment by sarbes** [comment]
Works fine on X11. I have yet to test it on GBM and Wayland, especially with a EGL_KHR_partial_update device.

File: xbmc/windowing/X11/GLContextEGL.cpp
Comment (neo1973) [inline xbmc/windowing/X11/GLContextEGL.cpp:554]: Is there no runtime check necessary to check if `EGL_BUFFER_AGE_EXT` is supported (like in `CEGLContextUtils::GetBufferAge()`)?

**Comment by neo1973** [comment]
I would love to runtime test this but when I set the dirty region mode to something other than 3 the GUI initialization fails with an `EGL_BAD_CONFIG`. Any idea why that is? This is on a Ryzen 5700u, EGL version is 1.5 and therefor `EGL_SWAP_BEHAVIOR_PRESERVED_BIT` should work if I understand correctly?

Log: https://paste.kodi.tv/ijawucekib.kodi

File: xbmc/windowing/X11/GLContextEGL.cpp
Comment (sarbes) [reply xbmc/windowing/X11/GLContextEGL.cpp:554]: I guess it will result in an EGL_BAD_ATTRIBUTE error, which might leave the age undefined. I'll add a check.

**Comment by sarbes** [comment]
> I would love to runtime test this but when I set the dirty region mode to something other than 3 the GUI initialization fails with an `EGL_BAD_CONFIG`. Any idea why that is? This is on a Ryzen 5700u, EGL version is 1.5 and therefor `EGL_SWAP_BEHAVIOR_PRESERVED_BIT` should work if I understand correctly?
> 
> Log: https://paste.kodi.tv/ijawucekib.kodi

Weird. Maybe we can't control it on Wayland? I've removed the Kodi side setup, it shouldn't be needed anyway. If we have EGL_BUFFER_PRESERVED behavior, it should work. EGL_BUFFER_DESTROYED would be better, from an performance perspective (no copy of the last frame).

**Comment by neo1973** [comment]
> Maybe we can't control it on Wayland?

It was the same on GBM for me.

Now it works but there are artifacts, especially with dirty region algorithm 2. Here a video, I had to film because on the screen capture it looks different for some reason:

https://github.com/xbmc/xbmc/assets/3134317/74701da7-ebc0-492e-ad7d-83e7d50ae1d9

With algo 1 there is only the flickering in the side bar like in the beginning of the video. The sliding in out of the page content isn't corrupted.

**Comment by sarbes** [comment]
The video doesn't work on my phone, but I guess we need to keep another frame.

**Comment by sarbes** [comment]
Rebased. @neo1973 I need a debug log to diagnose any issues, as there are three different code paths. EGL_KHR_partial_update, EGL_EXT_buffer_age, and just a plain hardcoded value.

**Comment by neo1973** [comment]
Hi @sarbes, here the log with the latest commits: https://paste.kodi.tv/isuvefebom.kodi

Screen corruption looks the same as in the video from https://github.com/xbmc/xbmc/pull/22998#issuecomment-2067807678

**Comment by sarbes** [comment]
@neo1973 I've made a double whammy.
1) EGL_EXT_buffer_age and EGL_KHR_partial_update were never queried, as they are EGL extensions, not EGL client extensions. 
2) The dirty regions should contain the content of all buffer age frames + the current frame.

It should work now, would you be so kind to test again? (I should really looking into setting up some more devices...)

**Comment by neo1973** [comment]
Yes, works now :+1:

File: xbmc/guilib/GUIWindowManager.cpp
Comment (sarbes) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: The buffer age can be 0. To quote EGL_EXT_buffer_age:
```
    This extension lets applications query the age of the back buffer
    contents for an EGL surface as the number of frames elapsed since
    the contents were most recently defined. The back buffer can
    either be reported as invalid (has an age of 0) or it may be
    reported to contain the contents from n frames prior to the
    current frame.
```
This is the case after initializing the frame buffers (on application start, viewport change, etc).

File: xbmc/utils/EGLUtils.cpp
Comment (sarbes) [reply xbmc/utils/EGLUtils.cpp:679]: Regions shouldn't be empty (https://github.com/xbmc/xbmc/blob/e0bd0a775938e756cbc6087578e4d0952c359e8f/xbmc/guilib/DirtyRegionTracker.cpp#L60), I'm removing the check.

File: xbmc/guilib/GUIWindowManager.cpp
Comment (neo1973) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: > This is the case after initializing the frame buffers (on application start, viewport change, etc).

OK, that means the `bufferAge` variable is always `>0` and the `if (bufferAge)` check is not needed.

Comment (fritsch) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: bufferAge can at any time become zero. One example is the mesa etnaviv driver. When it realizes that OpenGLES applications do not need all the performance it might reduce its internal buffer count from maybe 5 to 4, which would mean that more stages need to repaint. So whenever this happens, we have to get all three last blits (incase of triple buffer) on the screen.

Comment (fritsch) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: And the other way round. new buffer (in driver) means full repaint once.

Comment (sarbes) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: I've moved the `bufferAge + 1`, so the checks make sense. I've also limited the maximum tracked buffers, in case there is something up with the driver.

Comment (sarbes) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: Whoops, my second message got delayed. @fritsch, are you satisfied?

**Comment by sarbes** [comment]
I've squashed the previous commits.

File: xbmc/guilib/GUIWindowManager.cpp
Comment (fritsch) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: Sure. Thx.

Comment (fritsch) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: I don't think we need the 20 ;-) but safe is safe. Imagine the amount am DDR that needs.

Comment (sarbes) [reply xbmc/guilib/GUIWindowManager.cpp:1358]: The 20 is a bit much, but I have not touched the number. When debugging, I'm usually pinging between 1 and 5.

**Comment by sarbes** [comment]
@fritsch could you hit the green button? Or is there something else?

**Comment by fritsch** [comment]
I am not the maintainer here of this code. As @neo1973 has tested it, he should approve.

Thanks for your work.

**Review by neo1973 — APPROVED** [review]

File: xbmc/utils/EGLUtils.cpp
Comment (enen92) [inline xbmc/utils/EGLUtils.cpp:442]: c++ style cast?

File: xbmc/utils/EGLUtils.cpp
Comment (enen92) [inline xbmc/utils/EGLUtils.cpp:675]: emplace_back?

File: xbmc/windowing/tvos/WinSystemTVOS.h
Comment (enen92) [inline xbmc/windowing/tvos/WinSystemTVOS.h:54]: is it possible to make this magic number something meaningful (create some enum class)?

File: xbmc/windowing/X11/GLContextEGL.h
Comment (enen92) [inline xbmc/windowing/X11/GLContextEGL.h:38]: maybe make this return a bool instead and adapt the main logic?

File: xbmc/windowing/tvos/WinSystemTVOS.h
Comment (sarbes) [reply xbmc/windowing/tvos/WinSystemTVOS.h:54]: I've just moved the value. Previously, it was defined in the region tracker.

Rather than creating an enum, I would set it to NoOfBuffers(). But that's something for its own PR. Is that okay?

File: xbmc/utils/EGLUtils.cpp
Comment (sarbes) [reply xbmc/utils/EGLUtils.cpp:675]: That's not possible, is it? At least not without having a local `Rect` in the loop to emplace. Or so my understanding.

Does this make a difference?

File: xbmc/utils/EGLUtils.cpp
Comment (sarbes) [reply xbmc/utils/EGLUtils.cpp:442]: Sure.

File: xbmc/utils/EGLUtils.cpp
Comment (neo1973) [reply xbmc/utils/EGLUtils.cpp:675]: > That's not possible, is it? At least not without having a local Rect in the loop to emplace. Or so my understanding.

You are correct, you can't `emplace_back` a `std::array` even in C++20 where it supports `emplace_back` for aggregates.

File: xbmc/windowing/tvos/WinSystemTVOS.h
Comment (enen92) [reply xbmc/windowing/tvos/WinSystemTVOS.h:54]: Sure...

File: xbmc/windowing/X11/GLContextEGL.h
Comment (sarbes) [reply xbmc/windowing/X11/GLContextEGL.h]: Done that, but I think it is a bit confusing now.

**Comment by sarbes** [comment]
jenkins build this please

**Pull request merged**
